text/2017-11-05-deque.md

```diff
@@ -0,0 +1,151 @@
+    pub fn len(&self) -> usize;
+
+    /// Steals an element from the top of the deque.
+    pub fn steal(&self) -> Steal<T>;
+}
+
+impl<T> Clone for Stealer<T> { ... }
+
+/// Possible outcomes of a steal operation.
+pub enum Steal<T> {
+    /// The deque was empty.
+    Empty,
+    /// Some data was stolen.
+    Data(T),
+    /// Inconsistent state was encountered. Try again.
+    Inconsistent,
+}
+```
+
+An interesting difference from Coco's API is that a deque is now
+constructed using `Deque::new()` rather than using a global function returning
+a `(Worker<T>, Stealer<T>)` pair.
+Also, there are two ways of creating multiple stealers: you can either create each of them
+with `Deque::stealer` or by clone an existing one - whatever works best for you.
+
+Another addition is the `with_min_capacity` constructor. Deques dynamically grow and
+shrink as elements are inserted and removed. By specifying a large minimum capacity
+it is possible to reduce the frequency of reallocations.
+
+The `steal` method returns a `Steal<T>`. Instead of retrying on a failed CAS operation, the
+method returns immediately with `Steal::Inconsistent`. This gives schedulers
+fine control over what to do in case of contention.
+
+There is an open [pull request](https://github.com/crossbeam-rs/crossbeam-deque/pull/1)
+implemented according to this RFC.
+
+# Drawbacks
+
+None.
+
+# Alternatives
+
+Ultimately, hazard pointers would probably be a better fit for the deque than
+epoch-based garbage collection.
+
+In theory, epoch-based GC may leak memory indefinitely if a pinned thread is preempted,
+while hazard pointers give stricter guarantees. HP-based GC guarantees that there
+is always a bounded number of still-unreclaimed garbage objects.
+On the other hand, the advantage of epoch-based GC is that it allows fast
+traversal through linked data structures. The Chase-Lev deque isn't a linked data
+structure, so here we don't gain anything by choosing epoch-based GC over HP-based GC.
+
+Moreover, HP-based GC would allow us to force perfectly eager destruction of
+buffers: just like `Arc`, but without the overhead of contended reference counting.
+The idea is that when a thread wants to destroy a buffer, it would check
+all hazard pointers to see if it is still used. If there is a thread using it,
+we'd CAS the hazard pointer using the buffer and set it to null. When that thread
+notices that someone has set its hazard pointer to null, it would take over the
+responsibility of destroying the buffer. Then it would check all hazard pointers to see if
+anyone is still using it and continue in the same vein.
+
+At some point in the future, we should definitely experiment with HP-based
+garbage collection.
+
+# Unresolved questions
+
+When building `futures-pool`, **[@carllerche](https://github.com/carllerche)**
+wanted a few additional methods:
+
+1. [A method that steals more than one value at a time.](https://github.com/stjepang/coco/issues/11#issuecomment-339785208)
```

Comment:
Definitely interesting, but I agree that this isn't a requirement for moving forward and could be explored later.